File: src/main/java/com/vc/hard/L354.scala
package com.vc.hard

object L354 {
  def maxEnvelopes(arr: Array[Array[Int]]): Int = {
    if(arr.length == 0) return 0
    //[[2,3], [5,4], [6,4], [6,7]]
    object EOrdering extends Ordering[Array[Int]] {
      def compare(x: Array[Int], y: Array[Int]): Int = {
        val cmp = x(0).compareTo(y(0))
        if(cmp != 0) return cmp
        else return x(1).compareTo(y(1))
      }
    }
    scala.util.Sorting.quickSort(arr)(EOrdering)
    val n = arr.length
    val dp = new Array[Int](n)
    var i = 0
    var j = 0
    while(i < n) {
      j = 0
      dp(i) = 1
      while(j < i) {
        if(canInsert(j, i)) {
          dp(i) = Math.max(dp(i), dp(j) + 1)
        }
        j += 1
      }
      i += 1
    }
    def canInsert(small: Int, big: Int): Boolean = {
      arr(small)(0) < arr(big)(0) && arr(small)(1) < arr(big)(1)
    }
    dp.max
  }
}

File: src/main/java/com/vc/hard/L354.java
package com.vc.hard;

import java.util.*;

class L354 {
    public int maxEnvelopes(int[][] envelopes) {
        if(envelopes == null || envelopes.length == 0) return 0;

        Arrays.sort(envelopes, new Comparator<int[]>(){
            public int compare(int[] x, int[] y) {
                int cmp = Integer.compare(x[0], y[0]);
                if(cmp == 0) return Integer.compare(x[1], y[1]);
                return cmp;
            }
        });

        int n = envelopes.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        for(int i = 1; i < n; i++) {
            for(int j = 0; j < i; j++) {
                if(canInsert(envelopes[j], envelopes[i])) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
        }

        int max = 0;
        for(int i = 0; i < n; i++) {
            max = Math.max(dp[i], max);
        }
        return max;
    }

    private boolean canInsert(int[] x, int[] y) {
        return x[0] < y[0] && x[1] < y[1];
    }
}